README
DESCRIPTION

  kurenma  is a package that includes tools for configuration and control
  WireGuard tunnels by using PKI infrastructure (OpenSSL).
  Prepared debian packages can be found in the repository ppa:yds/bashlyk
  (sudo add-apt-repository ppa:yds/bashlyk).

COMMANDS

  kurenmactl
      management tool for kurenmad(8) daemon

  kurenmad
      daemon  for  automatic  configuration  and control of the WireGuard 
      tunnels

  kurenma-ssl-gen
      generate   CA   infrastructure   based   on   OpenSSL   and   x.509
      certificate/key pair for client and server peers

  kurenma-knockd-setup
      get   knock-sequence   of   port   numbers   from   the    server's
      X.509-certificate and write the appropriate section of the settings
      in the knockd.conf(5) configuration file. Only required for  server
      roles

  kurenma-firehol-update
      adds  the  IP-address to the white list of the firehol(1) firewall.
      It's designed to run from the service knockd(1) to allow  WireGuard
      client access



ChangeLog
Sun, 21 Jan 2018 13:33:47 +0400
	New upstream release
	Version 0.9
 - acceptance of naming convention for functions from the bashlyk version 1.91.XX
   that allows to use a clearer description of the configuration and code
 - add prevents for openssl hangups (watching), improved control of the SSL/TLS
   exchange, reduced time of the exchange, added handling of the [ips allowed]
   section for the client role
 - option 'pause between checks' also used in the server code for removing
   obsolete connections in the global cycle, disabled by default
 - added new option 'protocol' for secure channel transport setting. By default
   the initial handshake uses a version-flexible method which will negotiate the
   highest mutually supported protocol version
 - redesigned openssl command line, fixed format line, added a debugging message
   (level 6) to display the openssl command line
 - update-firehol - temporary fix for openssl error "gethostbyname failure" on
   the server side by  using fake names of the knocked IP-addresses in the
   /etc/hosts
 - improved logging
 - updated man and robodoc guides

client.kurenma.src:
 - udfSetupTunnel - 'ip link del dev ..' supress stderr
 - added checking SSL command stopping status
 - removed unused functions for extended control of the SSL exchange
 - udfGetTunnelSettings - watchdog forced stopped if not required
 - 'openssl s_client ...' - for ps ':' replaced by ' ' (!)
 - server availibity checked by ping for DTLS and netcat for other
   protocol

server.kurenma.src:
 - show subject and issuer from depth={0,1}
 - kurenma-server::handleRequest - removed deprecated code
 - used simple template for openssl command stopping
 - added functions kurenma-server::{checkConnections,inactivityConnections} for
   extended control of the connections


Wed, 01 Mar 2017 12:35:53 +0400
	New upstream release
	Version 0.8
 - configuration options become more clear through the use of bashlyk version
   1.89.700 with new INI implementation (option names with whitespaces):
        dev           -> tunnel name
        tsKeepalive   -> persistent keepalive
        tsPeerTimeout -> pause between checks
        bDefaultRoute -> default route
        port          -> listen port
        server        -> tunnel network
        pool          -> IP lease method
        host          -> server
        portAuth      -> port (moved to [ssl] section)
 - INI-options and variables can now be different
 - '{try,catch}-every-line' box replaced by 'try .. catch' implementation from
   bashlyk 1.89.689
 - fixed error of the DNS name resolving
 - improved code style and debian scripts - fixed lintian messages, updated
   descriptions
 - added special file VERSION for tracking version and revision from kurenmad
   and kurenmactl

server.kurenma.src:
 - fixed format error for generating remote peer properties
 - fixed condition for tunnel setup
 - delete the temporary file when the failure preparing the remote peer
   properties
 - 'connection' word replaced by 'exchange' for client interaction
   logging

kurenmad, setup-knockd:
 - use nonbreaked whitespace as leader char for display the configuration info

kurenmactl:
 - added variable _kurenma_pathLib

setup-knockd:
 - used 'INI::show' instead 'cat ..'
 - support knockd INI parser with use INI::settings..
 - split stop and start service by delay 2 sec

debian/postrm:
 - added missing variables

Makefile.am:
 - added target ..VERSION to syslib_DATA


Mon, 07 Nov 2016 15:56:10 +0400
	New upstream release
	Version 0.7
 - used SSL/TLS (s_client, s_server) interaction instead of SMIME for the
   initial key exchange
 - certificates for clients and servers are now different
 - added support for revoked certificates (-crl_check_all)
 - used trusted CA certificate for knock sequence generating from SHA1
   Fingerprint
 - changed some main configuration parameters:
   'ipServer' -> 'server' (CIDR ip/mask now)
   'ipClient' -> 'pool' ( 'dynamic' or 'commonname' )
 - udf{StopProcess,GetTimeInSec} removed and used from bashlyk version 1.87
 - updated tests and manuals

kurenmad:
 - udfSignal changes:
   - send "stopsignal" to pipe of openssl
   - kill all child 'sleep' command by udfStopProcess (bashlyk 1.87)
   - set _kurenma_sSignal='stop' earlier than stopped sleep
 - udfWaitRequest changes:
   - used return code of bash builtin command "read":
     - >128 - timeout
     - >0   - closed input
   - added support of s_server 'Q' command (shutdown) on event "stopsignal"
   - improved CN and data handling

server.kurenma.src:
 - openssl - used s_server instead smime
 - input wait each time increased by 10 seconds in the event of downtime
 - used tag 'SSL' for the openssl output logging
 - fixed dynamic leased IP range

client.kurenma.src:
 - openssl - used s_client instead smime
 - used trusted CA certificate for knock sequence getting from SHA1 Fingerprint
 - improved SSL files checking

kurenma.ini.example:
 - option 'ipServer' renamed to 'server' and now is the CIDR
 - option 'ipClient' renamed to 'pool'
 - section ssl:
   - parameter cn{Client,Server} replaced by 'cn'
   - added parameters 'ca', 'dh'

gen-ssl-kurenma:
 - 'crt' option replaced by 'server' and 'client' for separate generation of
   certificates by using the parameter "nsCertType"
 - for role 'server' generated Diffie Hellman parameters

setup-knockd:
 - used trusted CA certificate for knock sequence generating from SHA1
   Fingerprint

kurenmactl:
 - udfStopProcess is runned if PID of service exist


Tue, 18 Oct 2016 11:12:06 +0400
	New upstream release
	Version 0.6
 - reduced cpu usage by stopping of daemon with signal
 - simplified and unified the output some error messages
 - kurenma.logrotated - enabled restart of daemon
 - improved installation/removal of debian package

 kurenmactl:
 - improved stopping of daemon
 - kurenma.ini configuration not required
 - stopping of netcat (nc) not required

 kurenmad:
 - added global variable $_kurenma_sSignal as status of signal

src/wgclient.sh
#!/bin/bash
#
# $Id: wgclient.sh 10 2016-07-25 01:34:53+04:00 toor $
#
_bashlyk=kurenma . bashlyk
#
#
#

udfMain() {

	DEBUGLEVEL=4

	local a cnClient cnServer dev fn fnKey fnTmp fnServerCrt fnClientCrt fnClientKey host i ini path pathKey pathCrt port s

	path=/etc/kurenma
	ini=${path}/client.kurenma.ini
	pathCrt=${path}/ssl/public
	pathKey=${path}/ssl/private

	udfThrowOnCommandNotFound echo ip grep knock openssl ping printf sed tee wg

	[[ $UID == 0 ]] || eval $( udfOnError throw  iErrorNotPermitted "You must be root to run this." )

	udfMakeTemp fnKey
	udfMakeTemp fnTmp

	udfIni $ini ':dev;host;port ssl:cnServer;cnClient'

	: ${dev:=wg0}
	: ${i:=128}

	udfDebug 2 "configuration:" && udfShowVariable dev host port cnServer cnClient
	udfThrowOnEmptyVariable dev host port cnServer cnClient

	fnServerCrt=${pathCrt}/${cnServer}.crt
	fnClientCrt=${pathCrt}/${cnClient}.crt
	fnClientKey=${pathKey}/${cnClient}.key

	s=$( grep Subject: $fnClientCrt | sed -re "s/.*CN=(.*)\/email.*/\1/" )

	if [[ $s != $cnClient ]]; then

		eval $( udfOnError throw iErrorNotValidArgument "invalid cert file $fnClientCrt" )

	fi

	if [[ -f $fnClientCrt && -f $fnClientKey  && -f $fnServerCrt ]]; then


		udfDebug 3 && printf "client auth info:\n\tremote public key\t- %s\n\tlocal private key\t- %s\n\tlocal Common Name \t- %s\n" "$fnServerCrt" "$fnClientKey" "$cnClient"

	else

		eval $( udfOnError throw iErrorNoSuchFileOrDir "$fnClientCrt and/or $fnClientKey and/or $fnServerCrt" )

	fi

	## TODO send knocks and wait for destination port availibity ( nc -z )
	udfDebug 2 "check server availibity:"
	while true; do

		if echo "${i}%8" | bc | grep '^0$' >/dev/null; then

			knock $host 22025 23501 37565 && echo -n "!"

		fi
		sleep 1

		nc -w 8 -z ${host} ${port} 2>/dev/null && break
		echo -n "."
		i=$((i-1))
		(( $i > 0 )) || eval $( udfOnError throw iErrorNotPermitted "${host}:${port}" )

	done
	udfDebug 2 "ok."
	sleep 4

	s="$( wg genkey | tee $fnKey | wg pubkey | tee $fnTmp | udfEcho - $cnClient | openssl smime -encrypt -aes256 -outform PEM $fnServerCrt | nc -w 64 -i 16 $host $port | openssl smime -decrypt -inform PEM -inkey $fnClientKey | tr -d '\r' )"

	udfDebug 4 && printf "\nclient wirequard keys:\n\tprivate\t- %s\n\tpublic\t- %s\n\n" "$(< $fnKey)" "$(< $fnTmp)"

	udfDebug 3 "server answer: $s"

	[[ $s =~ ^OK ]] || eval $( udfOnError throw iErrorNotValidArgument "bad answer from ${host}:${port} $s" )


	a=( ${s//:/ } )

	ip link del dev $dev 2>/dev/null || true
	ip link add dev $dev type wireguard
	wg set $dev private-key $fnKey peer "${a[1]}" allowed-ips 0.0.0.0/0 endpoint "${host}:${a[2]}"
	ip address add "${a[3]}"/24 dev $dev
	ip link set up dev $dev

	ping -t 1 -c 3 -q ${a[4]} >/dev/null 2>&1

	eval set -- $(_ sArg)

	if [[ "$1" == "default-route" ]]; then

		host="$(wg show $dev endpoints | sed -n 's/.*\t\(.*\):.*/\1/p')"
		echo "host for route to wg = $host"
		echo "route info for ${host}:"
		ip route get $host
		host="$( ip route get $host | sed '/ via [0-9]\{1,3\}\.[0-9]\{1,3\}\.[0-9]\{1,3\}\.[0-9]\{1,3\}/{s/^\(.* via [0-9]\{1,3\}\.[0-9]\{1,3\}\.[0-9]\{1,3\}\.[0-9]\{1,3\}\).*/\1/}' | head -n 1 )"
		echo "new route $host"
		ip route add $host 2>/dev/null || true
		ip route add 0/1 dev $dev
		ip route add 128/1 dev $dev

	fi

	udfDebug 1 "client configuration done."

	udfDebug 3 && {

		printf "\n\nWireguard interface %s info:\n-----------------------------\n\n" "$dev"
		wg

		printf "\n\nWireguard interface %s configuration:\n--------------------------------------\n\n" "$dev"
		wg showconf $dev

		printf "\n----\n"

	} >&2

	wg showconf $dev > ${path}/${cnClient}_${cnServer}.${dev}.conf
	chmod 0600 ${path}/${cnClient}_${cnServer}.${dev}.conf

}
#
#
#
udfMain
#


src/wgserver.sh
#!/bin/bash
#
# $Id: wgserver.sh 10 2016-07-25 01:34:53+04:00 toor $
#
_bashlyk=kurenma . bashlyk
#
#
#
udfMakeFifo() {

	[[ -n "$1" ]] || eval $( udfOnError iErrorEmptyOrMissingArgument )
	udfMakeTemp $1 path=/var/run
	rm -f ${!1}
	mkfifo ${!1}

}
#
udfWaitRequest() {

	local -r iMax=64
	local i=0 s csv
	#
	while read s; do

		udfDebug 5 "input: $s ${i}" >&2
		csv+="${s}:"
		i=$((i+1))

		if (( $i > $iMax )); then

			echo "Warn: too many lines received, abort .." >&2
			break

		fi

		[[ $s =~ ^-----END.PKCS7-----$ ]] && break

	done

	if [[ $csv =~ ^-----BEGIN.PKCS7-----.*-----END.PKCS7-----:$ ]]; then

		udfHandleRequest "$csv"

	else

		echo "unexpected.."

	fi
}
#
udfHandleRequest() {

	## TODO udfThrowOnEmptyArgument
	[[ -n "$1" ]] || eval $( udfOnError retecho iErrorEmptyOrMissingArgument )

	udfThrowOnEmptyVariable fnServerKey ipServer dev pathCrt keepalive

	local fn s keyPeer fnClientCrt cnClient ipClient IFS

	udfMakeTemp fn

	IFS=':'
	for s in $1; do

		echo $s

	done | openssl smime -decrypt -inform PEM -inkey $fnServerKey | tr -d '\r' > $fn
	IFS=$' \t\n'

	udfDebug 5 && {

		printf "\nreceived data:\n"
		cat $fn

	} >&2

	if [[ $( wc -l < $fn ) != 3 ]]; then

		echo "Error - unexpected format"
		eval $( udfOnError retwarn iErrorNotValidArgument "untrusted input data" )

	fi

	keyPeer="$( tail -n +3 $fn )"
	s=$( head -n 1 $fn )
	rm -f $fn

	for fn in $( ls $pathCrt ); do

		if [[ -f $pathCrt/$fn && $fn =~ .crt ]]; then

			cnClient="$( grep 'Subject:' $fnClientCrt | sed -re "s/.*CN=(.*)\/email.*/\1/" )"

		else

			continue

		fi

		[[ $s == $cnClient ]] || continue

		fnClientCrt=${pathCrt}/$fn
		break

	done

	if [[ -z "$fnClientCrt" ]]; then

		echo "Error - unexpected request"
		eval $( udfOnError retecho iErrorNotPermitted "client certificate not found" )

	fi

	udfDebug 4 && printf "\nremote peer info:\n\tSerial No.\t- %s(%s)\n\twg public key\t- %s\n" "$s" "${fnClientCrt##*/}" "$keyPeer" >&2

	if [[ $(wg show $dev | grep peer | wc -l) -ge 253 ]]; then

		wg set $dev peer $(wg show $dev latest-handshakes | sort -k 2 -b -n | head -n 1 | cut -f 1) remove

	fi

	ipClient=$( s="$( wg show $dev allowed-ips )"; for ((i=2; i<=254; i++)); do ip="${ipServer%.*}.$i"; [[ $s != *${ip}/32* ]] && echo $ip && break; done )

	udfMakeTemp fn
	if wg set $dev peer "$keyPeer" allowed-ips "${ipClient}/32" persistent-keepalive $keepalive 2>$fn; then

		echo "OK:$( wg show $dev private-key | wg pubkey ):$( wg show $dev listen-port ):${ipClient}:${ipServer}"

	else

		echo "Error $(< $fn) $?"

	fi | openssl smime -encrypt -outform PEM $fnClientCrt

	udfDebug 1 "server configuration updated" >&2

	udfDebug 3 && {

		printf "\n\nWireguard interface %s info:\n-----------------------------\n\n" "$dev"
		wg

		printf "\n\nWireguard interface %s configuration:\n--------------------------------------\n\n" "$dev"
		wg showconf $dev

		printf "\n----\n"

	} >&2

	wg showconf $dev > $path/${cnServer}.${dev}.conf
	chmod 0600 ${path}/${cnServer}.${dev}.conf

}
#
udfMain() {

	DEBUGLEVEL=4

	[[ $UID == 0 ]] || eval $( udfOnError throw iErrorNotPermitted "You must be root to run this." )

	local dev fn fnServerKey fnTmp ini ip ipServer ipClient keyPeer path pathKey pathCrt keepalive s

	udfThrowOnCommandNotFound cut echo grep ip kill nc openssl ps sort wg

	udfExitIfAlreadyStarted

	path=/etc/kurenma
	pathCrt=${path}/ssl/public
	pathKey=${path}/ssl/private
	ini=${path}/server.kurenma.ini

	udfIni $ini ':dev;port;portAuth;ipServer;keepalive ssl:cnServer'

	: ${dev:=wg0}
	: ${ipServer:=192.168.29.1}
	: ${port:=12912}
	: ${portAuth:=42912}
	udfIsNumber $keepalive && [[ $keepalive -gt 10 && $keepalive -lt 3600 ]] || keepalive=333

	udfDebug 2 && udfShowVariable dev ipServer port portAuth cnServer
	udfThrowOnEmptyVariable cnServer

	if [[ -z "$( wg | grep $dev )" || -z "$( ip addr show $dev | grep $ipServer )" ]]; then

		ip link del dev $dev 2>/dev/null
		ip link add dev $dev type wireguard
		ip address add ${ipServer}/24 dev $dev
		wg set $dev private-key <(wg genkey) listen-port $port
		ip link set up dev $dev

	fi

	fnServerKey=${pathKey}/${cnServer}.key

	[[ -n $fnServerKey && -f $fnServerKey ]] || eval $(udfOnError exitecho iErrorNoSuchFileOrDir "$fnServerKey")

	s=$( ps -C nc -o pid= -o args= | grep -w "$portAuth" | cut -f 1 -d' ' )
	[[ -n "$s" ]] && kill -9 $s

	udfMakeFifo fnTmp
	cat $fnTmp | udfWaitRequest | nc -l -k -v $portAuth > $fnTmp

}
#
#
#
udfMain
#


src/wgmakepem.sh
#!/bin/bash
#
# $Id: wgmakepem.sh 1 2016-07-10 18:56:46+04:00 toor $
#
_bashlyk=openssl . bashlyk
#
#
#
udfMain() {

	udfThrowOnCommandNotFound echo chmod mkdir openssl touch wg

	local path pathSSL conf confTemplate

	path=/etc/wg/ssl
	pathPub=${path}/public
	pathPrv=${path}/private
	conf=${path}/wg.ssl
	confTemplate=${path}/openssl.cnf.template
#
	mkdir -p ${path}/{public,private,certs,newcerts,crl}
	chmod 0710 $pathPrv

	[[ -f $confTemplate ]] && cp -fv $confTemplate $conf

	echo $(date "+%s") > ${path}/serial

	rm -fv ${path}/index.txt
	touch ${path}/index.txt
# 
	openssl req -nodes -new -x509 -keyout ${pathPrv}/cakey.pem -out ${pathPub}/cacert.pem -days 3650 -config $conf -verbose
	openssl dhparam -out ${pathPub}/dh1024.pem 1024
#
	wg genkey | tee ${pathPrv}/private.wg.key | wg pubkey > ${pathPub}/public.wg.key
#
	chmod 0600 ${pathPrv}/*

}
#
#
#
udfMain
#


src/curenmakecrt.sh
#!/bin/bash
#
# $Id: curenmakecrt.sh 20 2016-08-02 12:55:59+04:00 toor $
#
_bashlyk=kurenma . bashlyk
#
#
#
_defLocal() {

	local s S

	for s in ${*//[;,]/ }; do

		udfIsValidVariable $s || eval $( udfOnError throw iErrorNonValidVariable "$s" )
		S+="$s "

	done

	echo "local ${S% *}"

}
#
udfMain() {

	eval set -- $( _ sArg )

	[[ -n "$1" ]] || eval $( udfOnError throw iErrorEmptyOrMissingArgument )

	udfThrowOnCommandNotFound chmod cp openssl sed

	local conf fnTemplate path pathCnf pathCrt pathCsr pathKey s csv='C;ST;L;O;OU;CN;EA'
	eval $(_defLocal $csv)

	path=/etc/kurenma/ssl
	pathCsr=${path}/csr
	pathCrt=${path}/public
	pathKey=${path}/private
	pathCnf=${path}/configs
	conf=${pathCnf}/${1}.kurenma.ssl
	confDN=${path}/dn.kurenma.ini
	fnTemplate=${path}/openssl.cnf.template
#
	for s in "$path" "$pathKey" "$pathCrt" "$pathCsr" "$pathCnf"; do

		[[ -d "$s" ]] || eval $( udfOnError throw iErrorNoSuchFileOrDir "$s" )

	done

	for s in "$confDN" "$fnTemplate"; do

		[[ -f $s ]] || eval $( udfOnError throw iErrorNoSuchFileOrDir "$s" )

	done

	cp -f $fnTemplate $conf

	udfIni $confDN ":${csv}"
#
	: ${C:=XX}
	: ${ST:=State}
	: ${L:=Locality}
	: ${O:=Organizational}
	: ${OU:=OrganizationalUnit}
	: ${EA:=postmaster@$1}

	OU=${1%%.*}
	CN=$1

	for s in ${csv//[;,]/ }; do

		sed -i -r -e "s/%${s}%/${!s}/ig" $conf

	done
#
	openssl req -new -nodes -keyout ${pathKey}/${1}.key -out ${pathCsr}/${1}.csr -config $conf -verbose
	openssl ca  -in ${pathCsr}/${1}.csr -out ${pathCrt}/${1}.crt -config $conf -verbose
#
	chmod 0600 ${pathKey}/${1}.key

}
#
#
#
udfMain
#


src/curenmakepem.sh
#!/bin/bash
#
# $Id: curenmakepem.sh 20 2016-08-02 15:04:23+04:00 toor $
#
_bashlyk=kurenma . bashlyk
#
#
#
_defLocal() {

	local s S

	for s in ${*//[;,]/ }; do

		udfIsValidVariable $s || eval $( udfOnError throw iErrorNonValidVariable "$s" )
		S+="$s "

	done

	echo "local ${S% *}"

}
#
udfMain() {

	udfThrowOnCommandNotFound echo chmod mkdir openssl rm sed touch wg

	local conf confDN fnTemplate path pathCnf pathCrt pathKey s csv='C;ST;L;O;OU;CN;EA'
	eval $(_defLocal $csv)
#
	path=/etc/kurenma/ssl
	pathCrt=${path}/public
	pathKey=${path}/private
	pathCnf=${path}/configs
	conf=${pathCnf}/kurenma.ssl
	confDN=${path}/dn.kurenma.ini
	fnTemplate=${path}/openssl.cnf.template
#
	mkdir -p ${path}/{public,private,configs,certs,newcerts,crl,csr}
	chmod 0710 $pathKey

	[[ -f $fnTemplate ]] && cp -fv $fnTemplate $conf

	echo $(date "+%s") > ${path}/serial

	rm -f ${path}/index.txt
	touch ${path}/index.txt
#
	udfIni $confDN ":${csv}"
#
	: ${C:=XX}
	: ${ST:=State}
	: ${L:=Locality}
	: ${O:=Organizational}
	: ${OU:=OrganizationalUnit}
	: ${CN:=CommonName}
	: ${EA:=email@domain.name}

	for s in ${csv//[;,]/ }; do

		sed -i -r -e "s/%${s}%/${!s}/ig" $conf

	done
#
	openssl req -nodes -new -x509 -keyout ${pathKey}/cakey.pem -out ${pathCrt}/cacert.pem -days 3650 -config $conf -verbose
	#openssl dhparam -out ${pathCrt}/dh2048.pem 2048
#
	wg genkey | tee ${pathKey}/private.wg.key | wg pubkey > ${pathCrt}/public.wg.key
#
	chmod 0600 ${pathKey}/*

}
#
#
#
udfMain
#


src/ssl-server-socat.sh
#!/bin/bash
#
# $Id: ssl-server-socat.sh 112 2016-10-20 17:24:56+04:00 toor $"
#
_bashlyk=kurenma _bashlyk_log=nouse . bashlyk
#
udfMain() {

	DEBUGLEVEL=3

	[[ $UID == 0 ]] || eval $( udfOnError throw NotPermitted "You must be root to run this." )

	eval set -- $( _ sArg )

	udfThrowOnEmptyOrMissingArgument $@

	local cn fn fnDH fnKey fnCert path pathCert

	cn=$1
	path=/etc/kurenma/ssl
	pathCert=${path}/certs
	fnDH=${path}/public/dh2048.pem
	fnCert=${path}/public/${1}.crt
	fnKey=${path}/private/${1}.key

	udfDebug 3 "openssl s_server -cert $fnCert -key $fnKey -accept 443 -CApath $pathCert -Verify 2 -crl_check_all -dhparam $fnDH"

	udfMakeTemp fn type=pipe

	udfWaitRequest < $fn | socat -d -d -d -d openssl-listen:443,cert=${fnCert},key=${fnKey},capath=${pathCert},dhparam=${fnDH} > $fn 2>&1

}
#
#
#
udfWaitRequest() {

    local s
    while read s; do

	if [[ $s =~ ^[0-9]+/[0-9]+/[0-9]+.[0-9]+:[0-9]+:[0-9]+.socat.*$ ]]; then

	    if [[ $s =~ subject:.*CN=.*$ ]]; then

		s=${s##*/CN=}
		s=${s%/*}

		cd /etc/kurenma/ssl

		openssl verify -crl_check -CApath certs public/${s}.crt || exit 1

	    fi

	else

	    echo "data $s"

	fi

    done

}
#
#
#
udfMain
#


src/ssl-server-openssl.sh
#!/bin/bash
#
# $Id: ssl-server-openssl.sh 114 2016-10-21 16:12:24+04:00 toor $"
#
#_bashlyk=kurenma _bashlyk_log=nouse . bashlyk
_bashlyk=kurenma . bashlyk
#
#
#
#
udmResetCounters() {

	bRevoked=0
	cn=''
	tsStart=0
	tsWait=60

}
#
udmCloseConnection() {

	udfDebug 3 "client $cn connection closed"

	udmResetCounters

	echo "connection down"
	sleep 1
	echo "q"

}
#
udmRevoked() {

	udfDebug 3 "client $cn cert revoked"
	echo "client cert revoked"

	udmCloseConnection

}
#
udmLimitConnectionTime() {

	udfIsNumber $1 || eval $( udfOnError return InvalidArgument "$1 - number expected" )
	udfIsNumber $tsStart || eval $( udfOnError return InvalidArgument "$tsStart - number expected" )
	(( tsStart > 0 )) || return 0
	[[ -n "$cn" ]] || eval $( udfOnError return EmptyArgument "cn" )

	local i=$(( $(date "+%s") - $tsStart ))

	(( i > 16 )) && udfDebug 3 "client $cn connection time is ${i}sec"

	if (( $i > $1 )); then

		udfDebug 3 "client $cn connection down by timeout ${i}sec"
		echo "client $cn connection down by timeout"

		udmCloseConnection

	fi

}
#
udfWaitRequest() {

	local bRevoked=0 cn i s tsStart=0 tsWait=60 rc

	while true; do

		read -t $tsWait s
		rc=$?

		#udfDebug 3 "rc=$rc : $s"

		if   (( rc > 128 )); then

			(( tsStart == 0 )) && udfDebug 3 "wait ${tsWait}sec"
			(( bRevoked > 0 )) && udmRevoked

		elif (( rc > 0 )); then

			udfDebug 3 "input closed"
			break

		elif [[ $s =~ ^signalStop$ ]]; then

			udfDebug 3 "SSL server down"

			sleep 0.2
			echo Q
			sleep 0.2
			break

		elif [[ $s =~ ^verify.*revoked.* ]]; then

			udfDebug 3 "warn: $cn revoked certificate"
			bRevoked=1
			tsWait=2

		elif [[ $s =~ depth=0.*CN.* ]]; then

			s="${s##*, CN = }"
			s="${s%,*}"

			[[ "$s" == "$cn" ]] && continue || cn=$s

			udfDebug 3 "client $cn connection start"

			tsStart=$(date "+%s")
			tsWait=4

		elif [[ $s =~ client-data.* ]]; then

			if (( bRevoked > 0 )); then

				udfDebug 3 "client data ignored"

			else

				udfDebug 3 "client data:${s##*client-data}"
				udfHandleRequest $cn ${s##*client-data}
				sleep 1.1
				udmCloseConnection

			fi


		elif [[ $s =~ ^(depth|verify|CIPHER|ERROR|CONNECTION|ACCEPT|subject|issuer) ]]; then

			udfDebug 3 "log $s"

		fi

		udmLimitConnectionTime 32

	done

}
#
udfHandleRequest() {

	udfThrowOnEmptyOrMissingArgument "$1" "$2"
	udfThrowOnEmptyVariable ipClient ipServer dev pathDat pathIni tsKeepalive tsPeerTimeout

	local aCIDR cnClient fn fnLeased i ip peer s

	cnClient=$1
	peer=$2

	fnLeased=${pathDat}/${cnServer}.${dev}.leased

	udfDebug 3 && \
		printf -- "\nremote peer info:\n\tClient Common Name\t- %s\n\tPublic key for %s\t- %s\n" \
			"$cnClient" "$dev" "${peer:0:10}<skipped>" >&2

	if [[ $(wg show $dev | grep peer | wc -l) -ge 253 ]]; then

		wg set $dev peer $(wg show $dev latest-handshakes | sort -k 2 -b -n | head -n 1 | cut -f 1) remove

	fi

	for s in $( wg show $dev latest-handshakes | tr '\t' ':' | xargs ); do

		if udfIsNumber ${s##*:}; then

			#ignore too low timestamps
			(( ${s##*:} < 1800 )) && continue

			i=$(( $(date +%s) - ${s##*:} ))

			if (( i >= tsPeerTimeout )); then

				wg set $dev peer ${s%%:*} remove
				udfDebug 2 "peer ${s:0:10}<skipped> is removed due to inactivity over an ${i}s .. $?"

			fi

		else

			eval $( udfOnError warn InvalidArgument "${s##*:}" )

		fi

	done

	if [[ "$ipClient" == "commonname" ]]; then

		ipClient=$( udfGetValidIPsOnly $cnClient )
		udfDebug 5 "Leased IP by Common Name to $cnClient - $ipClient"

	fi

	if [[ -z "$ipClient" || $ipClient == "dynamic" ]]; then

		[[ -f $fnLeased ]] || touch $fnLeased

		ipClient=$( grep -P "^${cnClient}:\S+$" $fnLeased | tail -n 1 | cut -f 2 -d':' | xargs )
		ipClient=$( udfGetValidIPsOnly $ipClient )

	fi

	if [[ -z "$ipClient" ]]; then

		s="$( wg show $dev allowed-ips )"

		for ((i=${ipServer##*.}; i<=254; i++)); do

			ip="${ipServer%.*}.$i";

			if [[ $s =~ ${ip}/32 ]]; then

				continue

			else

				ipClient=$( udfGetValidIPsOnly $ip )
				break

			fi

		done

	fi

	if [[ -z "$ipClient" ]]; then

		echo "not allocated IP to the client"
		eval $( udfOnError retecho EmptyOrMissingArgument "not allocated IP to the client" )

	fi

	aCIDR=$( udfGetValidCIDR ${ipClient}/32 ${ipsAllowed//[;,]/} )

	udfDebug 5 "allowed ips: $aCIDR"

	udfMakeTemp fn
	if wg set $dev peer "$peer" allowed-ips "${aCIDR// /,}" persistent-keepalive $tsKeepalive 2>$fn; then

		echo "OK:$( wg show $dev private-key | wg pubkey ):$( wg show $dev listen-port ):${ipClient}:${ipServer}"

		for s in $( wg show $dev peers ); do

			grep ":${s:0:10}$" $fnLeased

		done > $fn
		mv -f $fn $fnLeased
		echo "${cnClient}:${ipClient}:$(date +%s):${peer:0:10}" >> $fnLeased

	else

		echo "Error $(< $fn) $?"

	fi

	for s in $( wg show $dev allowed-ips | grep none | cut -f 1 | xargs ); do

		wg set $dev peer $s remove
		udfDebug 3 "${dev}: peer ${s:0:10}<skipped> is removed as unused .. $?"

	done

	udfDebug 1 "server configuration updated."
	udfDebug 5 && udfShowPeerInfo >&2

	wg showconf $dev > ${pathIni}/${cnServer}.${dev}.conf
	chmod 0600 ${pathIni}/${cnServer}.${dev}.conf

}
#
udfSetupTunnel() {

	udfThrowOnEmptyVariable dev ipServer port

	udfDebug 3 "Init WireGuard device $dev"

	try-every-line

		ip link del dev $dev 2>/dev/null || true
		ip link add dev $dev type wireguard
		ip address add ${ipServer}/24 dev $dev
		wg set $dev private-key <(wg genkey) listen-port $port
		ip link set up dev $dev

	catch-every-line

	udfDebug 3 "init new server pub key: $( wg show $dev private-key | wg pubkey | head -c 10 )<skipped>"

	return 0

}
#
udfService() {

	DEBUGLEVEL=3

	[[ $UID == 0 ]] || eval $( udfOnError throw NotPermitted "You must be root to run this." )

	udfThrowOnCommandNotFound cut kill ps sort
	udfThrowOnEmptyVariable cnServer dev ipServer pathCA pathCrt pathKey port portAuth

	local cmdSSL fn fnDH fnKey fnCrt

	if [[ -z "$( wg | grep $dev )" || -z "$( ip addr show $dev | grep $ipServer )" ]]; then

		udfSetupTunnel

	fi

	fnDH=${pathCrt}/dh2048.pem
	fnCrt=${pathCrt}/${cnServer}.crt
	fnKey=${pathKey}/${cnServer}.key
	cmdSSL="openssl s_server -cert $fnCrt -key $fnKey -accept $portAuth -CApath $pathCA -Verify 2 -crl_check_all -dhparam $fnDH"
	udfStopProcess "$cmdSSL"

	while true; do

		udfDebug 3 "Daemon for WireGuard clients (re)started.."

		udfMakeTemp fn type=pipe

		export _kurenma_control_channel=$fn
		udfWaitRequest < $fn | $cmdSSL > $fn 2>&1 &
		udfAddPid2Clean $!
		wait $!
		rm -f $fn

		udfWaitSignal 0.2 || break


	done

	udfStopProcess "$cmdSSL"

}
#
#
#
